Core/include/Acts/Utilities/detail/Subspace.hpp

```diff
@@ -0,0 +1,368 @@
+  /// implementation detail. Users should treat the return type as a generic
+  /// container whose elements are convertible to `size_t`.
+  constexpr const std::array<uint8_t, kSize>& indices() const { return m_axes; }
+
+  /// Check if the given axis index in the full space is part of the subspace.
+  constexpr bool contains(size_t index) const {
+    bool isContained = false;
+    for (auto a : m_axes) {
+      isContained |= (a == index);
+    }
+    return isContained;
+  }
+
+  /// Project a full vector into the subspace.
+  ///
+  /// @tparam fullspace_vector_t Veector type in the full space
+  /// @param full Vector in the full space
+  /// @return Subspace vector w/ just the configured axis components
+  ///
+  /// @note Always returns a column vector regardless of the input
+  template <typename fullspace_vector_t>
+  auto projectVector(const Eigen::MatrixBase<fullspace_vector_t>& full) const
+      -> SubspaceVectorFor<fullspace_vector_t> {
+    EIGEN_STATIC_ASSERT_VECTOR_SPECIFIC_SIZE(fullspace_vector_t, kFullSize);
+
+    SubspaceVectorFor<fullspace_vector_t> sub;
+    for (auto i = 0u; i < kSize; ++i) {
+      sub[i] = full[m_axes[i]];
+    }
+    return sub;
+  }
+
+  /// Expand a subspace vector into the full space.
+  ///
+  /// @tparam subspace_vector_t Subspace vector type
+  /// @param sub Subspace vector
+  /// @return Vector in the full space w/ zeros for non-subspace components
+  ///
+  /// @note Always returns a column vector regardless of the input
+  template <typename subspace_vector_t>
+  auto expandVector(const Eigen::MatrixBase<subspace_vector_t>& sub) const
+      -> FullspaceVectorFor<subspace_vector_t> {
+    EIGEN_STATIC_ASSERT_VECTOR_SPECIFIC_SIZE(subspace_vector_t, kSize);
+
+    FullspaceVectorFor<subspace_vector_t> full;
+    full.setZero();
+    for (auto i = 0u; i < kSize; ++i) {
+      full[m_axes[i]] = sub[i];
+    }
+    return full;
+  }
+
+  /// Projection matrix that maps from the full space into the subspace.
+  ///
+  /// @tparam scalar_t Scalar type for the projection matrix
+  template <typename scalar_t>
+  auto projector() const -> ProjectionMatrix<scalar_t> {
+    ProjectionMatrix<scalar_t> proj;
+    proj.setZero();
+    for (auto i = 0u; i < kSize; ++i) {
+      proj(i, m_axes[i]) = 1;
+    }
+    return proj;
+  }
+
+  /// Expansion matrix that maps from the subspace into the full space.
+  ///
+  /// @tparam scalar_t Scalar type of the generated expansion matrix
+  template <typename scalar_t>
+  auto expander() const -> ExpansionMatrix<scalar_t> {
+    ExpansionMatrix<scalar_t> expn;
+    expn.setZero();
+    for (auto i = 0u; i < kSize; ++i) {
+      expn(m_axes[i], i) = 1;
+    }
+    return expn;
+  }
+};
+
+/// Variable-size subspace representation.
+///
+/// @tparam kFullSize Size of the full vector space
+/// @tparam kMaxSize Maximum size of the subspace
+template <size_t kFullSize, size_t kMaxSize = kFullSize>
+class VariableSizeSubspace {
```

Comment:
Maybe I'm missing the point (still on my phone), but can you explain the use case of this?